R/utilities-noise-generation.R

```diff
@@ -0,0 +1,187 @@
+#'
+#' @param n An integer giving number of variates needed.
+#' @param seed An integer indicating a seed for R's PNRG, defaults to \code{NULL}.
+#' 
+#' @return Random numeric vector of length \code{n} containing values between
+#'    zero and one.
+#'
+#' Draws secure random variates from the uniform distribution through \code{openssl}.
+#' If a seed is provided, the \code{runif} function is used to draw the random variates.
+#' @examples
+#' 
+#' uniform_secure <- dpUnif(n=1000)
+#' uniform_repeatable <- dpUnif(n=1, seed=75436)
+#' @seealso \code{\link{dpNoise}}
+#' @rdname dpUnif
+#' @export
+dpUnif <- function(n, seed=NULL) {
+    if (!is.null(seed)) {
+        set.seed(seed)
+        return(runif(n))
+    }
+    return(openssl::rand_num(n))
+}
+
+
+#' Differentially Private Noise Generator
+#' 
+#' Compile noise from a cryptographically secure random variates to achieve
+#'    differentially private statistics.
+#'
+#' @param n An integer giving number of variates needed.
+#' @param scale Numeric, the scale for the distribution.
+#' @param dist A character specifying the distribution from which to draw the 
+#'    noise.
+#' @param shape An integer giving the shape parameter for the gamma
+#'    distribution. Default to \code{NULL}.
+#' @param seed An integer indicating a seed for R's PNRG, defaults 
+#'    to \code{NULL}.
+#'
+#' @return Cryptographically secure noise vector or matrix.
+#' @examples
+#'
+#' laplace_noise <- dpNoise(n=1000, scale=1, dist='laplace')
+#' gaussian_noise <- dpNoise(n=1000, scale=1, dist='gaussian')
+#' laplace_noise_repeatable <- dpNoise(n=1, scale=1, dist='laplace', seed=96845)
+#' @seealso \code{\link{dpUnif}}
+#' @rdname dpNoise
+#' @export
+dpNoise <- function(n, scale, dist, shape=NULL, seed=NULL) {
+    u <- dpUnif(n, seed)
+    if (dist == 'laplace') {
+        return(qlap(u, b=scale))
+    } else if (dist == 'gaussian') {
+        return(qnorm(u, sd=scale))
+    } else if (dist == 'gamma') {
+        return(qgamma(u, scale=scale, shape=shape))
+    } else {
+        stop(sprintf('Distribution "%s" not understood', dist))
+    }
+}
+
+
+#' Random draw from Laplace distribution
+#'
+#' @param mu numeric, center of the distribution
+#' @param b numeric, spread
+#' @param size integer, number of draws
+#' 
+#' @return Random draws from Laplace distribution
+#' @examples
+#' 
+#' rlap(size=1000)
+#' @export
+rlap = function(mu=0, b=1, size=1) {
+    p <- runif(size) - 0.5
+    draws <- mu - b * sgn(p) * log(1 - 2 * abs(p))
+    return(draws)
+}
+
+
+#' Probability density for Laplace distribution
+#'
+#' @param x numeric, value
+#' @param mu numeric, center of the distribution
+#' @param b numeric, spread
+#' 
+#' @return Density for elements of x
+#' @examples
+#' 
+#' x <- seq(-3, 3, length.out=61)
+#' dlap(x)
+#' @export
+dlap <- function(x, mu=0, b=1) {
+    dens <- 0.5 * b * exp(-1 * abs(x - mu) / b)
+    return(dens)
+}
+
+
+#' LaPlace Cumulative Distribution Function
```

Comment:
Sorry for the nitpick:
```suggestion
#' Laplace Cumulative Distribution Function
```

Reply:
begin nitpicky is good, I appreciate the attention to detail